### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by OseSem**
This PR adds the initial commit for the template adding everything.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    rev: 5.13.2
+    hooks:
+      - id: isort
+        args: ["--profile", "black"]
+
+
+  - repo: https://github.com/psf/black
+    rev: 24.2.0
+    hooks:
+      - id: black
```
Comment (elenakrittik) [inline .pre-commit-config.yaml:22]: I'd suggest replacing isort with ruff's `I` rule group and black with ruff's `ruff format` command

File: README.md [hunk 2]
```diff
@@ -1,2 +1,70 @@
...
+2. Installing the dependencies: (Remove `--without dev` if you want to install dev dependencies)
+```bash
+python3 -m pip install poetry
+poetry install
+```
+3. Create a `.env` file in the root directory and add the following:
+```env
+TOKEN=YOUR_BOT_TOKEN
+```
+4. Update version and title in `pyproject.toml` and `src/__init__.py`.
```
Comment (elenakrittik) [inline README.md:30]: There should be a step suggesting to replace the license with whatever would be appropriate for the user (but still including a notice somewhere to comply with MIT's terms). Also, this bullet should also suggest to edit `authors` and `repository`

File: README.md [hunk 3]
```diff
@@ -1,2 +1,70 @@
...
+5. Run the bot: (Run `exit` to exit the shell.`)
+```bash
+poetry shell
+python3 main.py
+```
+
+## Examples
+### Making a cog:
```
Comment (elenakrittik) [inline README.md:38]: Plugins should be called plugins, not cogs

File: README.md [hunk 4]
```diff
@@ -1,2 +1,70 @@
...
+We use the `disnake.ext.plugins` module to create a plugin for the bot. This defeates the purpose of having to make a class for each Cog and is way simplified.
```
Comment (elenakrittik) [inline README.md:39]:
```suggestion
This template uses [`disnake.ext.plugins`](https://github.com/DisnakeCommunity/disnake-ext-plugins) to nicely split bot's functionality into multiple files. Plugins are conceptually similar to cogs, but offer a simplified interface and don't rely on various hacks that cogs use internally.
```

File: README.md [hunk 5]
```diff
@@ -1,2 +1,70 @@
...
+async def example_ping(inter: disnake.CommandInteraction):
+    """Ping, Pong! [USAGE: /example ping]."""
+    await inter.response.send_message("Pong!")
+
+setup, teardown = plugin.create_extension_handlers()
+
+```
+
+### Using localization:
+Our template utilizated disnake's built-in i18n with our own small function that formats them with the arguments provided.
```
Comment (elenakrittik) [inline README.md:62]: Shameless plug: you may be interested in https://github.com/disnake-era/disnake-ext-fluent

No push though!

File: main.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        Fix this by adding the required privileged intents for your bot inside of:
+        | https://discord.com/developers/applications/{bot.user.id}/bot
+        """
+        logger.critical(msg)
+        if not bot.is_closed():
+            await bot.close()
+
+
+if __name__ == "__main__":
+    sys.exit(asyncio.run(main()))
```
Comment (elenakrittik) [inline main.py:64]: Unless i misunderstand the purpose behind all of this, you can effectively do the same this way:
```py
import asyncio
import contextlib

async def main():
    bot = ...

    await bot.start()

if __name__ == "__main__":
    with contextlib.supress(asyncio.CancelledError, KeyboardInterrupt):
        asyncio.run(main())
```

File: pyproject.toml [hunk 7]
```diff
@@ -5,13 +5,17 @@ description = "A basic template for a discord bot with disnake."
...
+package-mode = false
 
 [tool.poetry.dependencies]
 python = "^3.10"
 disnake = "^2.9.1"
-disnake-ext-plugins = "^0.0.0a0"
+disnake-ext-plugins = {git = "https://github.com/disnakecommunity/disnake-ext-plugins", rev = "master"}
 coloredlogs = "^15.0.1"
 psutil = "^5.9.8"
+helply = {git = "https://github.com/dlchamp/helply", rev = "main"}
```
Comment (elenakrittik) [inline pyproject.toml:18]: Same as above, this should be pinned to a specific commit

File: src/bot.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        """Execute when bot is connected to the Discord API."""
+
+    async def on_ready(self) -> None:
+        """Execute when bot is ready and cache is populated."""
+        msg = constants.generate_startup_table(bot_name=self.user.name, bot_id=self.user.id)
+        logger.info(f"\n{msg}")
+
+        self.loop_activities.start()
+
+    @tasks.loop(seconds=5)
```
Comment (elenakrittik) [inline src/bot.py:70]: Discord prohibits behaviour like this. At least set it to a few minutes

File: src/bot.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+                    type=constants.Client.activity_type,
+                ),
+                status=constants.Client.activity_status,
+            )
+        else:
+            logger.warning("There are no activities provided.")
+            await self.change_presence(activity=None, status=constants.Client.activity_status)
+            self.loop_activities.stop()
+
+    def load_extensions(self, path: str) -> None:
```
Comment (elenakrittik) [inline src/bot.py:86]: How is this different from the default load_extensions?

File: src/components/trash.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+from src.constants import Emoji
+
+TRASH_BUTTON_CUSTOM_ID = "TRASH:"
+
+
+class TrashButton(disnake.ui.Button[None]):
+    """A button that will delete the message when clicked.
+
+    This button uses low-level component listeners (@bot.listen('on_button_click')
```
Comment (elenakrittik) [inline src/components/trash.py:13]: Button classes cannot use or not use LLCs by themselves, only the caller decides

File: src/errors/__init__.py [hunk 11]
```diff
@@ -0,0 +1,7 @@
+from .base import BaseBotError
+from .http import GeneralHTTPError
```
Comment (elenakrittik) [inline src/errors/__init__.py:2]: these errors are small enough to not warrant their own files. even disnake with tens of errors keeps them in a single file

File: src/log.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
+import logging
+import logging.handlers
+from pathlib import Path
+
+import coloredlogs  # type: ignore[reportMissingTypeStubs]
```
Comment (elenakrittik) [inline src/log.py:5]: consider using loguru instead, no pressure though as i don't know how it compares to coloredlogs

File: src/log.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+# silence disnake's annoying info logger
+logging.getLogger("disnake").setLevel(logging.WARNING)
+
+logger = logging.getLogger()
+logger.info("Logging has been initialized")
+
+
+def get_logger(name: str) -> logging.Logger:
```
Comment (elenakrittik) [inline src/log.py:59]: this is a pretty much useless alias

File: src/util/http.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    return await response.text(encoding="utf-8")
+
+
+class Route:
+    """Handle route construction for HTTP requests."""
+
+    def __init__(self, method: str, url: str, **params: int | str | bool) -> None:
+        self.method = method
+
+        if params:
```
Comment (elenakrittik) [inline src/util/http.py:52]: use yarl for this

File: src/util/http.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
+from __future__ import annotations
```
Comment (elenakrittik) [inline src/util/http.py:1]: i'm not sure this is general or useful for most bots enough to include in the template

File: README.md [hunk 16]
```diff
@@ -1,2 +1,70 @@
...
+
+
+## Key Features
+- Neat packaging control.
+- Workflows for linting.
+- Basic bot setup.
+
+## Getting Started
+1. Clone the template ***[here](https://github.com/new?template_name=bot-template&template_owner=OseSem)***.
+2. Installing the dependencies: (Remove `--without dev` if you want to install dev dependencies)
```
[the remark this answers is not in the document]
Comment (OseSem) [reply README.md:21]: Oh yea I removed `--without dev` but forgot but forgot removing that.

File: README.md (hunk 3 above)
Comment (OseSem) [reply README.md:38]: Fixed.

File: src/errors/__init__.py (hunk 11 above)
Comment (OseSem) [reply src/errors/__init__.py:2]: The base is just 1 file and then there is another file housing the errors, but yea I will make it 1 file in the main `src` directory

File: src/log.py (hunk 12 above)
Comment (OseSem) [reply src/log.py:5]: I like using logging with coloredlogs as logging is built into python.

File: src/log.py (hunk 13 above)
Comment (OseSem) [reply src/log.py:59]: Tried with the suggestion you commented before and I got no output.

File: src/util/http.py (hunk 15 above)
Comment (OseSem) [reply src/util/http.py:1]: Then they can use it whenever they like.

File: src/bot.py (hunk 8 above)
Comment (OseSem) [reply src/bot.py:70]: Worked fine for me but I will change it.

File: src/bot.py (hunk 9 above)
Comment (OseSem) [reply src/bot.py:86]: No setup function error and logging.

File: src/util/http.py (hunk 14 above)
Comment (OseSem) [reply src/util/http.py:52]: For what part?

File: main.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
+import asyncio
+import os
+import signal
+import sys
+
+import disnake
+
+from src import constants, log
```
[the remark this answers is not in the document]
Comment (OseSem) [reply main.py:8]: If I named them `{bot_name}` like I did on my other projects the user would have to change the folder name + every import.

File: pyproject.toml (hunk 7 above)
Comment (OseSem) [reply pyproject.toml:18]: Both fixed.

File: src/bot.py (hunk 8 above)
Comment (OseSem) [reply src/bot.py:70]: Changed to 5 minutes.

File: src/errors/__init__.py (hunk 11 above)
Comment (OseSem) [reply src/errors/__init__.py:2]: Fixed.

File: README.md (hunk 2 above)
Comment (OseSem) [reply README.md:30]: Done.

File: .gitignore [hunk 18]
```diff
@@ -153,8 +153,20 @@ dmypy.json
 cython_debug/
 
 # PyCharm
-#  JetBrains specific template is maintained in a separate JetBrains.gitignore that can
+#  JetBrains specific template is maintainted in a separate JetBrains.gitignore that can
```
[the remark this answers is not in the document]
Comment (OseSem) [reply .gitignore:156]: Fixed.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (OseSem) [reply .pre-commit-config.yaml:22]: I like having isort

File: src/util/http.py (hunk 14 above)
Comment (elenakrittik) [reply src/util/http.py:52]: Everything related to manipulating URLs, like you do in the highlighted method

Comment (OseSem) [reply src/util/http.py:52]: So for the parameters? I don't get what you are saying as I never used yarl

Comment (elenakrittik) [reply src/util/http.py:52]: Yes, you are concatenating query parameters by hand, and if i was doing it i'd instead use a trusted library for that

Comment (OseSem) [reply src/util/http.py]: Should be fixed.

File: src/components/trash.py (hunk 10 above)
Comment (OseSem) [reply src/components/trash.py:13]: Should be fixed now.

**Pull request merged**